tests/bluetooth/audio/ascs/src/test_ase_state_transition_invalid.c

```diff
@@ -0,0 +1,624 @@
+
+	ase_id = test_ase_id_get(fixture->ase_src);
+	test_preamble_state_streaming(conn, ase_id, stream, &chan, true);
+
+	test_client_config_codec_expect_transition_error(conn, ase_id, ase_cp);
+	test_client_config_qos_expect_transition_error(conn, ase_id, ase_cp);
+	test_client_enable_expect_transition_error(conn, ase_id, ase_cp);
+	test_client_receiver_start_ready_expect_transition_error(conn, ase_id, ase_cp);
+	test_client_receiver_stop_ready_expect_transition_error(conn, ase_id, ase_cp);
+}
+
+ZTEST_F(test_ase_state_transition_invalid, test_client_source_state_disabling)
+{
+	const struct bt_gatt_attr *ase_cp = fixture->ase_cp;
+	struct bt_bap_stream *stream = &fixture->stream;
+	struct bt_conn *conn = &fixture->conn;
+	uint8_t ase_id;
+
+	Z_TEST_SKIP_IFNDEF(CONFIG_BT_ASCS_ASE_SRC);
+
+	ase_id = test_ase_id_get(fixture->ase_src);
+	test_preamble_state_disabling(conn, ase_id, stream);
+
+	test_client_config_codec_expect_transition_error(conn, ase_id, ase_cp);
+	test_client_config_qos_expect_transition_error(conn, ase_id, ase_cp);
+	test_client_enable_expect_transition_error(conn, ase_id, ase_cp);
+	test_client_receiver_start_ready_expect_transition_error(conn, ase_id, ase_cp);
+	test_client_disable_expect_transition_error(conn, ase_id, ase_cp);
+	test_client_update_metadata_expect_transition_error(conn, ase_id, ase_cp);
+}
+
+static void test_server_config_codec_expect_error(struct bt_bap_stream *stream)
+{
+	struct bt_codec codec = BT_CODEC_LC3_CONFIG_16_2(BT_AUDIO_LOCATION_FRONT_LEFT,
+							 BT_AUDIO_CONTEXT_TYPE_UNSPECIFIED);
+	int err;
+
+	err = bt_bap_stream_reconfig(stream, &codec);
+	zassert_false(err == 0, "bt_bap_stream_reconfig unexpected success");
+}
+
+static void test_server_receiver_start_ready_expect_error(struct bt_bap_stream *stream)
+{
+	int err;
+
+	err = bt_bap_stream_start(stream);
+	zassert_false(err == 0, "bt_bap_stream_start unexpected success");
+}
+
+static void test_server_disable_expect_error(struct bt_bap_stream *stream)
+{
+	int err;
+
+	err = bt_bap_stream_disable(stream);
+	zassert_false(err == 0, "bt_bap_stream_disable unexpected success");
+}
+
+static void test_server_update_metadata_expect_error(struct bt_bap_stream *stream)
+{
+	struct bt_codec_data meta[] = {
+		BT_CODEC_DATA(BT_AUDIO_METADATA_TYPE_STREAM_CONTEXT,
+			      (BT_AUDIO_CONTEXT_TYPE_RINGTONE & 0xFFU),
+			      ((BT_AUDIO_CONTEXT_TYPE_RINGTONE >> 8) & 0xFFU)),
+	};
+	int err;
+
+	err = bt_bap_stream_metadata(stream, meta, ARRAY_SIZE(meta));
+	zassert_false(err == 0, "bt_bap_stream_metadata unexpected success");
+}
+
+ZTEST_F(test_ase_state_transition_invalid, test_server_sink_state_codec_configured)
+{
+	struct bt_bap_stream *stream = &fixture->stream;
+	struct bt_conn *conn = &fixture->conn;
+	uint8_t ase_id;
+
+	Z_TEST_SKIP_IFNDEF(CONFIG_BT_ASCS_ASE_SNK);
+
+	ase_id = test_ase_id_get(fixture->ase_snk);
+	test_preamble_state_codec_configured(conn, ase_id, stream);
+
+	test_server_receiver_start_ready_expect_error(stream);
```

Comment:
Shouldnt we also have wrong_direction error for a receiver stop ready here? And that goes for the rest of the sinks as well. A faulty client could send these illegally (just as other faulty transitions).
Maybe its out of scope for this unit test. Im not sure exactly what error it would produce right now.

Reply:
Those are tests for invalid API usage. Currently, the `bt_bap_stream_stop` is guarded with `CONFIG_BT_BAP_UNICAST_CLIENT`

Reply:
Why are those tests for invalid API usage? If a remote device (a client) sends a receiver stop ready to a sink ASE, that's not an API thing, but an invalid request from a remote device

Reply:
This particular test case is testing invalid API usage. see the prefix `test_server_`